rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -261,7 +261,7 @@ class Publisher : public PublisherBase
     if (this->can_loan_messages()) {
       // we release the ownership from the rclpp::LoanedMessage instance
       // and let the middleware clean up the memory.
-      this->do_loaned_message_publish(loaned_msg.release());
+      this->do_loaned_message_publish(loaned_msg.release().get());
```

Comment:
can we change `do_loaned_message_publish` so it takes a `rclcpp::LoanedMessage` by value and also make `rclcpp::LoanedMessage` a **move only** type? (like `std::unique_ptr`).

IMO, it's more clear that ownership is being transferred in the following snippet:

```suggestion
      this->do_loaned_message_publish(std::move(loaned_msg));
```

The `release` method isn't needed anymore with that API change (it can be deleted or kept though).
IMO, breaking API here makes sense, this is almost not being used and the current API is a bit broken.

Reply:
by the way, `rclcpp::LoanedMessage` seems to be already a move only type.

Reply:
@ivanpauno 

> this->do_loaned_message_publish(std::move(loaned_msg));

I can update with it, but `release` still needs to be called inside `do_loaned_message_publish`.

> The `release` method isn't needed anymore

Sorry, I don't catch it. I thought that `release` is to set a flag to control whether to `return_loaned_message` in `rclcpp` or `rmw`, and we don't know if users will call the publish() method after creating `LoanedMessage`.

Reply:
I have confused some things because of miss-reading the current code, sorry!

> can we change do_loaned_message_publish so it takes a rclcpp::LoanedMessage by value and also make rclcpp::LoanedMessage a move only type? (like std::unique_ptr).

This is already the case!
`rclcpp::Publisher::publish(rclcpp::LoanedMessage &&)` takes ownership.

> I think the release method must be put in private section of LoanedMessage, and then use friend class Publisher<MessageT, AllocatorT> in LoanedMessage.

So the comment [here](https://github.com/ros2/rclcpp/pull/1305#issuecomment-708868358) wasn't an actual leak of memory, only a potential leak if the user calls `release` and doesn't publish the message?
IMO, that's fine.
The idea of calling `release()` is "give me the raw ptr and I will make sure to free it correctly", so returning a raw ptr is the correct thing to do.
If the user call release, they must either publish it, or check with `can_loan_message()` the correct destructor to call and do it manually.

Reply:
>  The idea of calling release() is "give me the raw ptr and I will make sure to free it correctly", so returning a raw ptr is the correct thing to do.

👍

This pull request changes this for `rclcpp::LoanedMessage::release` just because in the case that the message was not loaned the logic to free it properly is a bit complicated (explicitly call destructor and use correct allocator, which the user may have not specified explicitly originally). But they still need to handle the unique_ptr with care, either publishing it or checking if it needs to be returned the middleware, so in that sense it's like a raw pointer half the time.

> If the user call release, they must either publish it, or check with can_loan_message() the correct destructor to call and do it manually.

Yes, but in the case that `can_loan_message()` returns false, they should just let it go out of scope.